idaes/apps/caprese/estimator.py

```diff
@@ -0,0 +1,597 @@
+                if was_originally_active[comp]:
+                    comp.activate()
+
+    def initialize_actualmeasurements_at_t0(self):
+        t0 = self.sample_points[0]
+        for ind in self.MEASUREMENT_SET:
+            actmea_block = self.ACTUALMEASUREMENT_BLOCK[ind]
+            mea_block = self.MEASUREMENT_BLOCK[ind]
+            actmea_block.var[0].set_value(mea_block.var[0].value)  
+
+    def initialize_past_info_with_steady_state(self, 
+                                               desired_ss, 
+                                               ss_weights, 
+                                               solver):
+        '''
+        Before running MHE, we need to set up past information (measurements, states, etc.). 
+        Here we solve for a steady state and set it as past information.
+
+        Parameters
+        ----------
+        desired_ss : List of vardata, value tuples describing the
+                     desired steady stsates of these specified variables.
+        ss_weights : List of vardata, value tuples describing the
+                     weightss of these specified variables.
+
+        '''
+        self.add_steady_state_objective(desired_ss, ss_weights)
+        self.solve_steady_state(solver)
+
+        self.initialize_actualmeasurements_at_t0()
+        self.initialize_to_initial_conditions(ctype=(DiffVar, AlgVar, DerivVar, InputVar,))
+
+    def add_noise_minimize_objective(self,
+                                     model_disturbance_weights,
+                                     measurement_noise_weights,
+                                     givenform = "weight"):
+        '''
+        Set up the objective function for MHE. The given form can be either 'weight'
+        or 'variance'.
+
+        Parameters
+        ----------        
+        model_disturbance_weights: A list of vardata-value tuples describing 
+                                    the weight for model disturbance to be 
+                                    given to the objective term containing each variable.
+        measurement noise_weights: A list of vardata-value tuples describing 
+                                    the weight for measurement error to be 
+                                    given to the objective term containing each variable.
+        givenform: The form of given lists. It should be either "weight" (default)
+                    or "variance".
+        '''
+
+        if givenform not in ["weight", "variance"]:
+            raise RuntimeError("Wrong argument 'givenform' is given. "
+                   "Please assign either 'weight' or 'variance'.")
+
+        diff_id_list = [id(var[0]) for var in self.differential_vars]
+        for var, val in model_disturbance_weights:  
+            #Check whether the given variable is classified under diffvar
+            if id(var) not in diff_id_list:
+                raise RuntimeError(var.name, 
+                                   " is not a differential variable.")
+
```

Comment:
We should probably use t0 instead of 0 as an index for var, right? Also, can you make `diff_id_list` a set? This way the 
```python
id(var) not in diff_id_list
```
check will be O(1) instead of O(n).

For the future: I would like to change things to specify weights with CUIDs rather than vardata objects.

Reply:
Also, what if the user's `var` is not at `t0`? What if it is at some other point in time instead?

Reply:
Yeah, I think I still assume that the user gives the weights, variances, and setpoints at `t0`. We should decide a consistent way to let the user declare variables (maybe CUID?).